launcher_go/v2/docker/commands.go

```diff
@@ -52,3 +57,197 @@ func (r *DockerBuilder) Run() error {
+			cmd.Args = append(cmd.Args, "--env")
+			cmd.Args = append(cmd.Args, k)
+		}
+	}
+
+	// Order is important here, we add extra env after config's env to override anything set in env.
+	for _, e := range r.ExtraEnv {
+		cmd.Args = append(cmd.Args, "--env")
+		cmd.Args = append(cmd.Args, e)
+	}
+	for k, v := range r.Config.Labels {
+		cmd.Args = append(cmd.Args, "--label")
+		cmd.Args = append(cmd.Args, k+"="+v)
+	}
+	if !r.SkipPorts {
+		for _, v := range r.Config.Expose {
+			if strings.Contains(v, ":") {
+				cmd.Args = append(cmd.Args, "-p")
+				cmd.Args = append(cmd.Args, v)
+			} else {
+				cmd.Args = append(cmd.Args, "--expose")
+				cmd.Args = append(cmd.Args, v)
+			}
+		}
+	}
+	for _, v := range r.Config.Volumes {
+		cmd.Args = append(cmd.Args, "-v")
+		cmd.Args = append(cmd.Args, v.Volume.Host+":"+v.Volume.Guest)
+	}
+	for _, v := range r.Config.Links {
+		cmd.Args = append(cmd.Args, "--link")
+		cmd.Args = append(cmd.Args, v.Link.Name+":"+v.Link.Alias)
+	}
+	cmd.Args = append(cmd.Args, "--shm-size=512m")
+	if r.Rm {
+		cmd.Args = append(cmd.Args, "--rm")
+	}
+	if r.Restart {
+		cmd.Args = append(cmd.Args, "--restart=always")
+	} else {
+		cmd.Args = append(cmd.Args, "--restart=no")
+	}
+	if r.Detatch {
+		cmd.Args = append(cmd.Args, "-d")
+	}
+	cmd.Args = append(cmd.Args, "-i")
+
+	// Docker args override settings above
+	for _, f := range r.Config.DockerArgs() {
+		cmd.Args = append(cmd.Args, f)
+	}
+	for _, f := range r.ExtraFlags {
+		cmd.Args = append(cmd.Args, f)
+	}
+	cmd.Args = append(cmd.Args, "-h")
+	cmd.Args = append(cmd.Args, r.Hostname)
+	cmd.Args = append(cmd.Args, "--name")
+	cmd.Args = append(cmd.Args, r.ContainerId)
+	if len(r.CustomImage) > 0 {
+		cmd.Args = append(cmd.Args, r.CustomImage)
+	} else {
+		cmd.Args = append(cmd.Args, r.Config.RunImage())
+	}
+
+	for _, c := range r.Cmd {
+		cmd.Args = append(cmd.Args, c)
+	}
+
+	if !r.Detatch {
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		cmd.Stdin = r.Stdin
+	}
+	runner := utils.CmdRunner(cmd)
+	if r.DryRun {
+		fmt.Println(cmd)
+	} else {
+		if err := runner.Run(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type DockerPupsRunner struct {
+	Config         *config.Config
+	PupsArgs       string
+	SavedImageName string
+	ExtraEnv       []string
+	Ctx            *context.Context
+	ContainerId    string
+}
+
+func (r *DockerPupsRunner) Run() error {
+	rm := false
+	// remove : in case docker tag is blank, and use default latest tag
+	r.SavedImageName = strings.TrimRight(r.SavedImageName, ":")
+	if r.SavedImageName == "" {
+		rm = true
+	}
+	defer func(rm bool) {
+		if !rm {
+			time.Sleep(utils.CommitWait)
+			runCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			cmd := exec.CommandContext(runCtx, utils.DockerPath, "rm", "-f", r.ContainerId)
+			utils.CmdRunner(cmd).Run()
+			cancel()
+		}
+	}(rm)
+	commands := []string{"/bin/bash",
+		"-c",
+		"/usr/local/bin/pups --stdin " + r.PupsArgs}
```

Comment:
```suggestion
	commands := []string{
		"/bin/bash",
		"-c",
		"/usr/local/bin/pups --stdin " + r.PupsArgs,
	}
```

Small one for readability.